src/ossl/common.rs

```diff
@@ -8,60 +8,270 @@ use interface::*;
+        impl $name {
+            pub fn new(name: *const c_char) -> KResult<$name> {
+                let arg = unsafe {
+                    $in_fetch(get_libctx(), name, std::ptr::null_mut())
+                };
+                if arg.is_null() {
+                    return err_rv!(CKR_DEVICE_ERROR);
+                }
+                let ptr = unsafe {
+                    /* This is safe and requires no lifetimes because all _CTX_new()
+                     * functions in OpenSSL take a reference on the argument */
+                    $newctx(arg)
+                };
                 unsafe {
-                    $free(self.ptr);
+                    $in_free(arg);
+                }
+                if ptr.is_null() {
+                    return err_rv!(CKR_DEVICE_ERROR);
                 }
+                Ok($name { ptr: ptr })
             }
+
+            ptr_wrapper_returns!($ossl);
         }
 
-        unsafe impl Send for $name {}
-        unsafe impl Sync for $name {}
+        ptr_wrapper_tail!($name; $free);
     };
+
+    (fetch; $name:ident; $ossl:ident; $fetch:ident; $free:expr) => {
+        ptr_wrapper_struct!($name; $ossl);
+
+        impl $name {
+            pub fn new(name: *const c_char) -> KResult<$name> {
+                let ptr = unsafe {
+                    $fetch(get_libctx(), name, std::ptr::null_mut())
+                };
+                if ptr.is_null() {
+                    return err_rv!(CKR_DEVICE_ERROR);
+                }
+                Ok($name { ptr: ptr })
+            }
+
+            ptr_wrapper_returns!($ossl);
+        }
+
+        ptr_wrapper_tail!($name; $free);
+    }
+}
+
+ptr_wrapper!(ctx; EvpMdCtx; EVP_MD_CTX; EVP_MD_CTX_new; EVP_MD_CTX_free);
+ptr_wrapper!(ctx; EvpCipherCtx; EVP_CIPHER_CTX; EVP_CIPHER_CTX_new; EVP_CIPHER_CTX_free);
+
+ptr_wrapper!(ctx_from_name; EvpKdfCtx; EVP_KDF_CTX; EVP_KDF_CTX_new; EVP_KDF_CTX_free; EvpKdf; EVP_KDF_fetch; EVP_KDF_free);
+ptr_wrapper!(ctx_from_name; EvpMacCtx; EVP_MAC_CTX; EVP_MAC_CTX_new; EVP_MAC_CTX_free; EvpMac; EVP_MAC_fetch; EVP_MAC_free);
+
+ptr_wrapper!(fetch; EvpMd; EVP_MD; EVP_MD_fetch; EVP_MD_free);
+ptr_wrapper!(fetch; EvpCipher; EVP_CIPHER; EVP_CIPHER_fetch; EVP_CIPHER_free);
+
+#[derive(Debug)]
+pub struct EvpPkeyCtx {
+    ptr: *mut EVP_PKEY_CTX,
+}
+
+impl EvpPkeyCtx {
+    pub fn new(name: *const c_char) -> KResult<EvpPkeyCtx> {
+        let ptr = unsafe {
+            EVP_PKEY_CTX_new_from_name(get_libctx(), name, std::ptr::null())
+        };
+        if ptr.is_null() {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+        Ok(EvpPkeyCtx { ptr: ptr })
+    }
+
+    pub unsafe fn from_ptr(ptr: *mut EVP_PKEY_CTX) -> KResult<EvpPkeyCtx> {
+        if ptr.is_null() {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+        Ok(EvpPkeyCtx { ptr: ptr })
+    }
+
+    /*
+    pub fn empty() -> EvpPkeyCtx {
+        EvpPkeyCtx {
+            ptr: std::ptr::null_mut(),
+        }
+    }
+    */
```

Comment:
This was just stop gap, I asked on our rust channel and they suggested to use Cow, which does exactly what I need. I should have known, I used Cow previously, but somehow did not realize it could be easily used this way.
Replaced with simpler code

Reply:
Yeah empty() is quite unsafe, there were uses before I did various changes to limit the use, and the forgot to remove the comment, will drop all uses of empty in a fixup.